### ISSUE TITLE
自定义分类迁移任务中的loss和类别权重

### ISSUE THREAD

**Issue by haixiao990** [PaddlePaddle/PaddleFormers#810]
欢迎您对PaddleHub提出建议，非常感谢您对PaddleHub的贡献！
在留下您的建议时，辛苦您同步提供如下信息：
- 您想要增加什么新特性？
- 什么样的场景下需要该特性？

做分类任务时因为样本不均衡问题，想要选择不同的 loss，自定义类别权重或者样本权重，不知道有没有好的方式实现。

- 没有该特性的条件下，PaddleHub目前是否能间接满足该需求？

现在好像只能通过继承ClassifierTask，自组网实现。

- 增加该特性，PaddleHub可能需要变化的部分。
- 如果可以的话，简要描述下您的解决方案

后续希望可以通过参数配置的方式实现一些常用Loss的选择。

**Comment by Steffy-zxf** [comment]
你好！目前你可以通过继承相应的Task，改写_add_loss方法实现自定义loss。可以参考文档：
https://github.com/PaddlePaddle/PaddleHub/blob/release/v1.8/docs/tutorial/how_to_define_task.md

https://github.com/PaddlePaddle/PaddleHub/blob/release/v1.8/docs/tutorial/define_task_example.md

后续PaddleHub会继续升级，提供更加灵活的使用。敬请期待。

**Comment by haixiao990** [comment]
#自定义了一个采用focal loss 的分类任务，可以参考。

```
class ImageClassifierTask_FocalLoss(hub.ClassifierTask):
    def __init__(self, 
                 feature,
                 num_classes,
                 dataset=None,
                 feed_list=None,
                 data_reader=None,
                 startup_program=None,
                 config=None,
                 hidden_units=None,
                 metrics_choices="default",
                 class_weights = None,
                 gamma = 2.):
        
        self.class_weights = np.array(class_weights).astype('float32')
        self.gamma = gamma

        super(ImageClassifierTask_FocalLoss, self).__init__(
            feature=feature,
            num_classes=num_classes,
            dataset=dataset,
            data_reader=data_reader,
            feed_list=feed_list,
            startup_program=startup_program,
            config=config,
            hidden_units=hidden_units,
            metrics_choices=metrics_choices)

    def _add_loss(self):
        ce_loss = fluid.layers.cross_entropy(
            input=self.outputs[0], label=self.labels[0])
        one_hot_label = fluid.layers.one_hot(self.labels[0], depth=self.num_classes)
        alpha = fluid.layers.assign(self.class_weights)
        alpha = fluid.layers.matmul(one_hot_label, alpha)
        prob = one_hot_label*self.outputs[0]
        prob = fluid.layers.reduce_sum(prob, dim=-1)
        weight = fluid.layers.pow((1.0 - prob), self.gamma)
        focal_loss = alpha * weight * ce_loss
        focal_loss = alpha * ce_loss
        
        return fluid.layers.reduce_sum(focal_loss)
```

**Comment by haoyuying** [comment]
Since you haven't replied for more than 3 months, we have closed this issue/pr.
If the problem is not solved or there is a follow-up one, please reopen it at any time and we will continue to follow up.
由于您超过三个月未回复，我们将关闭这个issue/pr。
若问题未解决或有后续问题，请随时重新打开，我们会继续跟进。
